### PR TITLE
Support more pre-trained models for Transformer Encoder

### LINKED ISSUE #691: Support more pretrained models for Transformer Encoder
**Describe the feature**
 After this proposal users will be able to use all of the models from this list (https://huggingface.co/models) while using `BaseTransformerEncoder`. Not all of these models are intended for for information retrievial tasks, but some may be useful. Note that with the current implementation users are already having issues, see #662.

**Your proposal**
Use this [class](https://huggingface.co/transformers/model_doc/auto.html#automodelforpretraining) from transformers library.
The code is already implemented in my fork [here](https://github.com/festeh/jina/blob/feat-support-more-transformers/jina/executors/encoders/nlp/transformer.py), if you're interested, I can polish it a bit and make a PR.

Pros:
* Users will be able to load all available pre-trained models
* The number of lines of code is reduced, removed hardcoded imports

Cons:
* Hard-coding will now occur in other place. For each model one needs to decide how to get a sequence embedding from the embeddings of its tokens. I propose a simple heuristic: if a model is BERT-ish, take embedding of a first token, otherwise (XLNet, GPT-2 and other language models), take embedding of a last token. There's no guarantee that it would work for all models
* Did not test for all possible models, just considered existing corner cases (need to add a padding token for language model)

### PR DESCRIPTION
**Pull request by festeh**
Resolves #691

Main change:
* all models from https://huggingface.co/models are now available. It can be done via `AutoModelForPreTraining`

Other changes
* Fixed bug in `prune_mask`
* Added a heuristic for automatic creation sequence embedding from its tokens (if BERT-like, use first token, else last)
* Used `batch_encode_plus` to simplify batching
* `max_length` is set to `None` - so batches will be padded to largest sequence. I think user should determine this parameter based on the task
* Re-implemented saving logic. I got lost in fields like `raw_model_path`, `tmp_model_path`, so instead I just added `model_save_path` field. 
* Added unit tests that are using mock transformer model. The reason is that they can be run relatively (tokenizer is still downloaded) quickly as huge amount of time is saved for model initialization

Let's discuss if some of these changes are not desirable and try to come up with a better solution

### CODE REVIEW

File: jina/executors/encoders/helper.py [hunk 1]
```diff
@@ -48,8 +48,8 @@ def prune_mask(mask, cls_pos='head'):
         mask_row[0, 0] = 1
         result = np.tile(mask_row, (mask.shape[0], 1))
     elif cls_pos == 'tail':
-        for num_tokens in np.sum(mask, axis=1).tolist():
-            result[:, num_tokens - 1] = 1
+        for row_index, num_tokens in enumerate(np.sum(mask, axis=1).tolist()):
```
Comment (JoanFM) [inline jina/executors/encoders/helper.py:51]: nice catch!

**Comment by JoanFM** [comment]
Hello @festeh,

The CI won't pass because you have some bad commit messages. please refer to:
- [Commit Message Guideline for the First Time Contributor](https://github.com/jina-ai/jina/issues/553)
- [Contributing Guideline](https://github.com/jina-ai/jina/blob/master/CONTRIBUTING.md)
#553 :bookmark: Commit Message Guideline for the First Time Contributor
jina-ai/jina | Jun 21st | Added by GitHub
CONTRIBUTING.md
jina-ai/jina | Added by GitHub

**Comment by festeh** [comment]
Yeah, will fix commit message soon. Meanwhile, I discovered a bug with Tensorflow model, fixing it now...

**Comment by JoanFM** [comment]
The idea of using `AutoModel` and `AutoTokenizer` was also shared by #662. This PR may close this issue

**Comment by JoanFM** [comment]
recheckcla

File: jina/executors/encoders/nlp/transformer.py [hunk 2]
```diff
@@ -2,76 +2,92 @@
 __license__ = "Apache-2.0"
 
 import os
-
-import numpy as np
-
+from typing import Optional
```
Comment (JoanFM) [inline jina/executors/encoders/nlp/transformer.py:5]: numpy as np needed for type hint, lint is complaining about it

Comment (festeh) [reply jina/executors/encoders/nlp/transformer.py:5]: I think it's a bit hacky way, maybe explicitly use `noqa` statement if linter is an only issue? Or just remove np from type hint?

Comment (festeh) [reply jina/executors/encoders/nlp/transformer.py:5]: On the second thought, why not, let's return it back. Generally, there are a lot of place when `if False: import ...` are used. My comment was about this situation.

**Comment by JoanFM** [comment]
LGTM! Just not sure if `auto` should have preference over `cls`. Also we can adopt more mocks to improve our testing!
